### ISSUE TITLE
Job with concurrency key blocked when it shouldn't have, eventually was run after `duration` expired

### ISSUE THREAD

**Issue by leondmello** [rails/solid_queue#456]
We are observing that our jobs are sometimes unnecessarily blocked.
In the following example, for the same concurrency key, there are only 2 jobs. The first one completes immediately in just over a 100 milliseconds (`finished_at` value).
The second one takes over 30 minutes, we are guessing it is blocked by the concurrency control duration. The `scheduled_at` time of the second job is much after when the first job completes.
As per our understanding, the second job should have started immediately.
We confirmed that none of the jobs processes bounced during that time.

`...` indicates redacted data.

### Concurrency Config
```ruby
limits_concurrency to: 1, key: ->(action, params) do
  ...
end, duration: 30.minutes
```

### Queue Config
```yaml
dispatchers:
    - polling_interval: 1
      batch_size: 500
workers:
    - queues: [<queue of blocked job>, "*"]
      threads: 3
      processes: <%= ENV.fetch("JOB_CONCURRENCY", 1) %>
      polling_interval: 0.1
```

### Console queries
```ruby
SolidQueue::Job.where("DATE_PART('minutes', finished_at - scheduled_at) >= 30").count
=> 1

SolidQueue::Job.where(concurrency_key: '...').select(:active_job_id, :scheduled_at, :finished_at).order(:finished_at)
=>
[#<SolidQueue::Job:0x00007f062d151a50
  active_job_id: "60916418-e7a3-4f1e-9171-357b6e52c154",
  scheduled_at: "2024-12-16 01:20:09.441199000 +0000",
  finished_at: "2024-12-16 01:20:09.648985000 +0000",
  id: nil>,
 #<SolidQueue::Job:0x00007f062d151910
  active_job_id: "3c4f34da-9db7-430e-9cbe-c1214f98ccb9",
  scheduled_at: "2024-12-16 01:54:49.287841000 +0000",
  finished_at: "2024-12-16 02:24:50.264702000 +0000",
  id: nil>]

 ````
 
 ### Logs about the second job execution
 ```
I, [2024-12-16T02:24:50.154361 #84]  INFO -- : [ActiveJob] [...] [3c4f34da-9db7-430e-9cbe-c1214f98ccb9] Performing ... (Job ID: 3c4f34da-9db7-430e-9cbe-c1214f98ccb9) from SolidQueue(...) enqueued at 2024-12-16T01:54:49.287885003Z with arguments: ...
I, [2024-12-16T02:24:50.264573 #84]  INFO -- : [ActiveJob] [...] [3c4f34da-9db7-430e-9cbe-c1214f98ccb9] Performed ... (Job ID: 3c4f34da-9db7-430e-9cbe-c1214f98ccb9) from SolidQueue(...) in 110.35ms
  ```

**Comment by rosa** [comment]
> As per our understanding, the second job should have started immediately.

That's right. The first job should have unblocked the second job [here](https://github.com/rails/solid_queue/blob/019d7c7de42c10e5f2e86ab4aa771816c735f947/app/models/solid_queue/claimed_execution.rb#L70) 🤔 Would it be possible for you to enable Active Record logs at the debug level to see what's happening when the first job runs and tries to unblock it?

**Comment by leondmello** [comment]
That would be a lot of logging unfortunately I think. And this isn’t that common I think.

**Comment by leondmello** [comment]
I'll see if I can add it in our staging environment. My plan is to add an initializer which basically does
```ruby
module SolidQueue
  class Job
    def unblock_next_blocked_job
      ActiveRecord.verbose_query_logs = true
      super
    ensure
      ActiveRecord.verbose_query_logs = false
    end
  end
end
```

**Comment by leondmello** [comment]
@rosa Are debug logs prefixed with `SolidQueue::Semaphore` sufficient
Went with a different approach which right now allows debug logs prefixed with `SolidQueue::`
```ruby
class FilteredSqLogger < ActiveSupport::Logger
  def debug(progname = nil, &block)
    super if progname.match(/^\s+SolidQueue::/)
  end
end
```

**Comment by rosa** [comment]
I think you could try `SolidQueue::Semaphore` and `SolidQueue::BlockedExecution`. I think any clues would involve any of those records.

**Comment by leondmello** [comment]
Ended up going with `/^\s*SolidQueue::(?!Process)/` since `Process` was the most noisy. I will post an update to this issue once we see regression.

**Comment by rosa** [comment]
Ohhhh, this is super useful, @leondmello! I think I know what's going on, thanks to your logs. Looks like a tricky race condition.  Basically, what happens is that your first job finishes before the second one has been enqueued and blocked, not before enough to have unblocked the semaphore but _before enough_ to not see it blocked yet. Here's the first job finishing: 
```
finished_at: "2024-12-23 11:54:57.793273000 +0000"
```
Then, when that job finishes, it unblocks the semaphore, here: 
```
D, [2024-12-23T11:54:57.872605 #97] DEBUG --
```

Here's the second job enqueued: 
```
scheduled_at: "2024-12-23 11:54:57.832971000 +0000"
```
At that point the semaphore hasn't been updated yet, because it was updated at 11:54:57.872605, so it gets blocked. However, when this `SELECT` runs, the blocked execution mustn't have been created yet: 
```
D, [2024-12-23T11:54:57.879150 #97] DEBUG -- :   SolidQueue::BlockedExecution Load (5.8ms)  SELECT "solid_queue_blocked_executions".* FROM "solid_queue_blocked_executions" WHERE "solid_queue_blocked_executions"."concurrency_key" = '...' ORDER BY "solid_queue_blocked_executions"."priority" ASC, "solid_queue_blocked_executions"."job_id" ASC LIMIT 1 FOR UPDATE SKIP LOCKED
```

So it can't get unblocked, because it's not completely blocked yet. Ahh, tricky! 

What [isolation level](https://www.postgresql.org/docs/current/transaction-iso.html) are you running?

**Comment by leondmello** [comment]
> What [isolation level](https://www.postgresql.org/docs/current/transaction-iso.html) are you running?

We have this in our `application.rb`,

```ruby
    config.active_support.isolation_level = :fiber
```

But don't know the reason as to why that was added some time ago. Trying to find out.

**Comment by rosa** [comment]
@leondmello, no, no worries about that. I meant the transaction isolation level in your DB (I imagined you're using PostgreSQL because of the logs), but thinking more about it, it shouldn't make any difference, so no worries about it.  

It's a tricky one, I'm not sure right now about how to fix it but I'll keep thinking about it.

**Comment by rosa** [comment]
Thinking more about this and how to fix... Claiming the semaphore + enqueuing the job happens in a transaction, but the semaphore is not locked if it exists because [it's just checked](https://github.com/rails/solid_queue/blob/eedb4fe384471cc347e5918cb76608dc25fc4e1d/app/models/solid_queue/semaphore.rb#L43-L44), so it can be released while the transaction is ongoing, and you wouldn't know, regardless of whether  `READ COMMITTED` or `REPEATABLE READ` is used as isolation level (the most common isolation levels), because you'd just read that once in the beginning. 

I think, to fix this, we'd have to take a lock there, when checking the semaphore, and releasing it when the transaction is committed 🤔 However, this might introduce a new kind of deadlock I can't see right now (I've fixed a couple of deadlocks here in the past) so I need to think carefully about it.

**Comment by leondmello** [comment]
@rosa Any luck fixing this one?

**Comment by rosa** [comment]
Hey @leondmello, unfortunately not 😞  I've got an idea that might work, but I haven't been able to test it carefully as I think it's going to cause problems under certain load. I'm now focused on other different work so I won't have a lot of time in the next couple of months, but will try to get back to this soon.

**Comment by rosa** [comment]
It's been ages (over a year! 😱), but better late than never I suppose 😅 I came back and implemented the fix I had discussed in [this comment](https://github.com/rails/solid_queue/issues/456#issuecomment-2565473082).

I'm going to test it in our app (HEY) to see how it goes.

**Comment by leondmello** [comment]
Thanks !

**Comment by rosa** [comment]
We've been running this since Friday in production in HEY, and everything is going great, so I'm going to merge the PR and close this!

**Comment by rosa** [comment]
Fixed by #712

**Comment by leondmello** [comment]
Thanks @rosa, will try out [v1.3.2](https://github.com/rails/solid_queue/releases/tag/v1.3.2) with our instance
